src/client/datascience/jupyter/kernels/helpers.ts

```diff
@@ -278,3 +280,92 @@ export function isLocalLaunch(configuration: IConfigurationService) {
 
     return false;
 }
+
+export function findPreferredKernelIndex(
```

Comment:
This is the 'generic' algorithm for finding a preferred kernel. Shared by both local and remote. At least so far.

Reply:
Can we retain the preferred kernel.
My concern is if users open notebooks and it takes 10s to find kernels, then users cannot run a cell for 10s.
If instead we return the preferred kernel in 1s, they can run a kernel immediately.
E.g. we run into this often, we open a notebook & just run a cell, but have to wait because VSCode is fetching all of the kernels.

I understand we cannot fix this ritht now, however, with this change it becomes impossible as we have to wait for all kernels to be found first, even before we return the preferred kernel. I.e. we won't be able to make such improvements due to this.

Reply:
We already wait for all kernels to be returned as far as I can tell. The current code calls ```getKernelForLocalJupyterConnection``` to try and find the preferred kernel. That in turn calls into the kernelFinder to find a matching kernel. That in turn then enumerates all of the kernels.

Reply:
E.g. when searching for exact match, i do'nt think we need to go through all the kernels, either we find it or we don't & if we dont, then we pick the default , which would then require ALL kernels.

Reply:
Never mind, we can look at the telemetry & see how bad this is.

Reply:
> getKernelForLocalJupyterConnection to try and find the preferred kernel. That in turn calls into the kernelFinder to find a matching kernel. That in turn then enumerates all of the kernels.

Agreed.

Just don't like the fact that users have to wait for all kernels to be discovered for then to run a notebook. Feels like its a serious bottle neck, no point creating an issue either as we don't have the telemetry